src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,527 @@
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setBuildSuccessful(currentBuildSuccessful);
+
+      blockingStub.finalizeBuild(builder.build());
+
+      stateAcrossBuilds.previousBuildId = stateAcrossBuilds.currentBuildId;
+      stateAcrossBuilds.currentBuildId = null;
+      currentOriginalOutputPath = null;
+      currentRelativeOutputPath = null;
+    }
+  }
+
+  @Override
+  public void finalizeAction(Action action, MetadataHandler metadataHandler)
+      throws IOException, EnvironmentalExecException, InterruptedException {
+    // TODO: Would this be the right place to call into the remote
+    // output service to check whether any I/O errors occurred? If so,
+    // we should likely let createActionFileSystem() call into the
+    // remote output service to start capturing I/O errors.
+  }
+
+  private String fixupExecRootPath(PathFragment path) {
+    return path.relativeTo(currentRelativeOutputPath).toString();
+  }
+
+  private class GrpcBatchStat implements BatchStat {
+    private abstract class DumbFileStatus implements FileStatusWithDigest {
+      @Override
+      public long getSize() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+
+      @Override
+      public long getLastModifiedTime() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+
+      @Override
+      public long getLastChangeTime() throws IOException {
+        return -1;
+      }
+
+      @Override
+      public long getNodeId() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+
+      @Override
+      public byte[] getDigest() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+    }
+
+    private class RegularFileStatus extends DumbFileStatus {
+      private final long size;
+      private final byte[] digest;
+
+      RegularFileStatus(long size, byte[] digest) {
+        this.size = size;
+        this.digest = digest;
+      }
+
+      @Override
+      public boolean isFile() {
+        return true;
+      }
+
+      @Override
+      public boolean isDirectory() {
+        return false;
+      }
+
+      @Override
+      public boolean isSymbolicLink() {
+        return false;
+      }
+
+      @Override
+      public boolean isSpecialFile() {
+        return false;
+      }
+
+      @Override
+      public long getSize() throws IOException {
+        return size;
+      }
+
+      @Override
+      public byte[] getDigest() throws IOException {
+        return digest;
+      }
+    }
+
+    private class SymlinkFileStatus extends DumbFileStatus {
+      @Override
+      public boolean isFile() {
+        return false;
+      }
+
+      @Override
+      public boolean isDirectory() {
+        return false;
+      }
+
+      @Override
+      public boolean isSymbolicLink() {
+        return true;
+      }
+
+      @Override
```

Comment:
These can go to the base class. BTW you can consider giving the base class a boolean/enum indicating the type, so you don't need to override is subclasses.

Reply:
Moved to base class and made them static.

Considering that in the latest version of this PR, directories also have a property of their own (i.e., the last modified time), I think there is less of an incentive to add a boolean/enum to indicate the type. Feel free to reopen if I'm mistaken!